src/main/java/org/galatea/starter/entrypoint/StockPriceRestController.java

```diff
@@ -0,0 +1,156 @@
+      try {
+        final boolean hasMissingDays = stockDataList.isEmpty()
+            || DateTimeUtils.missingDays(startOfRange, startOfToday, stockDataList);
+        if (hasMissingDays) {
+          //the size for a compact call is 100 days worth of data, so only order full if needed.
+          final AvResponse response = avService.getAlphaVantageStockData(symbol, days > 100);
+
+          StockDay.updateAll(stockDataList, stockId, response);
+          repo.saveAll(stockDataList);
+
+          cacheStatus = "missing days in cache were added during this request.";
+        } else {
+          //filter list to any days that have incomplete stock data, excluding today
+          //ie - were acquired mid-day on a business day
+          final List<StockDay> requireUpdates =
+              stockDataList.stream().filter(s ->
+                  DateTimeUtils.duringBusinessHours(s.getUpdateTime())
+                      && s.getTradeDate().toLocalDate().isBefore(startOfToday)
+              ).collect(Collectors.toList());
+
+          //if there are any days with incomplete data, get a new fresh list and update them.
+          if (!requireUpdates.isEmpty()) {
+
+            final AvResponse response = avService.getAlphaVantageStockData(symbol, days > 100);
+
+            StockDay.updateAll(requireUpdates, stockId, response);
+            repo.saveAll(requireUpdates);
+
+            //otherwise, if the stock market is still open (or was open the last
+            //time we checked), get new data for today.
+            cacheStatus = "cache data was added / updated during this request.";
+          } else {
+            final StockDay latest = stockDataList.get(0);
+            //if there's a request from today, and today's request
+            // is from business hours, then we have to update the data
+            if (!latest.getTradeDate().toLocalDate().isBefore(startOfToday) && DateTimeUtils
+                .duringBusinessHours(latest.getUpdateTime())) {
+              final StockDay refresh = avService.getStockPricesForRange(symbol, false).get(0);
+
+              latest.update(refresh);
+              repo.save(latest);
+
+              cacheStatus =
+                  "data for today has been updated as of " + latest.getUpdateTime() + ".";
+            } else {
+              cacheStatus = "all data in this request has come from the local cache.";
+
+            }
+          }
+
+        }
+        metaData.addMessage("cache", cacheStatus);
+        metaData.addMessage("days", "there were ", days - stockDataList.size(),
+            " non-business day(s) in the requested range.");
+      } catch (StockSymbolNotFoundException ex) {
+        metaData
+            .addMessage("error", "there was no stock with the symbol ", ex.getSymbol(),
+                " found.");
+        stockDataList = Collections.emptyList();
+      }
+      //filter out any remaining days before the start date
+      stockDataList =

```

Comment:
✔ was able to clean this up so that rather than filtering it, sorting it, and reversing it in separate steps, I can just return the filtered & sorted-in-reverse list without assigning it to a variable.